libcxx/test/std/iterators/iterator.primitives/range.iter.ops/range.iter.ops.advance/iterator_count_sentinel.pass.cpp

```diff
@@ -213,6 +221,20 @@ constexpr bool test() {
     assert(i == iota_iterator{INT_MIN+1});
   }
 
+  // Check that we don't do an unneeded bounds check when decrementing a
```

Comment:
Question: can you briefly explain why this case is special?

Reply:
The existing tests for the "backwards" case seem to exclude bidirectional iterators: <https://github.com/llvm/llvm-project/blob/1d296ec226903a09de0215e05d852b2b80712855/libcxx/test/std/iterators/iterator.primitives/range.iter.ops/range.iter.ops.advance/iterator_count_sentinel.pass.cpp#L191-L199>

So the tests didn't cover the part of `std::ranges::advance` where the iterator is decremented "one by one".

Reply:
I'm not sure why the comment says that being a sized sentinel is a requirement -- at least in the [current draft](http://eel.is/c++draft/iterators#range.iter.op.advance-5) of the Standard, all I see is a requirement that the iterator and the sentinel are the same type. IIUC, if the comment were true, the new test case would actually be UB (because it calls `advance` with a negative `n` and a bidirectional iterator).

Assuming the comment is not correct, IMO the `!sized_sentinel_for` check creates more confusion than clarity. The bidirectional iterator concept doesn't require the type to be a sized sentinel for itself, and in tests we use "minimal" iterator types (types that only support the bare minimum operations required to satisfy the concept). So I think this would be clearer if we simply test with a bidirectional iterator and omit the `static_assert`s and the comment.

Reply:
Yes, that comment does not seem to be correct.

My intention with the `!sized_sentinel_for` check was to _not_ fall into this case: <https://eel.is/c++draft/iterators#range.iter.op.advance-6.1>, but instead test the other one: <https://eel.is/c++draft/iterators#range.iter.op.advance-6.2>.

But it should actually be possible to remove the new, "special case" test and instead improve the generic `check_backward` tests, right? Because as you mentioned, being `sized_sentinel_for<It, It>` is not actually a precondition for "`advance` negative `n`". In an improved `check_backward` test, one would only have to ensure the actual preconditions:

- `[bound, i)` is a range
- that `It` is a `std::bidirectional_iterator`
- that `S` (the sentinel type) is the same as `It`

Then, one could instantiate `check_backward` with different iterator types that either model `sized_sentinel_for<It, It>` or don't. This way it would be possible to get coverage for both ["range.iter.op.advance-6.1"](<https://eel.is/c++draft/iterators#range.iter.op.advance-6.1>) and ["range.iter.op.advance-6.2"](<https://eel.is/c++draft/iterators#range.iter.op.advance-6.2>). Does this sound sensible to you?

Reply:
OK, I've removed that misleading comment for the `check_backward` tests and adapted them to take bidirectional iterators. Then, I could also remove my "special case" test I had added, since the `check_backward` tests now cover the "decrement one by one" case as well.

Reply:
@ldionne Just wanted to double-check with you regarding the comment in case we're missing something.